package/mediacenter-osmc/patches/all-104-support-for-hardware-keyboards.patch

```diff
@@ -0,0 +1,333 @@
++    if (m_keyMods & XBMCKMOD_CTRL) keyMapTable = keyMapTable|K_CTRLTAB;
++    if (useKeyboard)
+     {
+-      int keyMapValue;
+-      if (devt.key.keysym.mod & (XBMCKMOD_SHIFT | XBMCKMOD_CAPS)) keyMapValue = entry.shift;
+-      else if (devt.key.keysym.mod & XBMCKMOD_ALT) keyMapValue = entry.alt;
+-      else if (devt.key.keysym.mod & XBMCKMOD_META) keyMapValue = entry.altShift;
+-      else keyMapValue = entry.base;
++      if (m_kbMods & XBMCKMOD_RALT) keyMapTable = keyMapTable|K_ALTTAB;
++      if (m_kbMods & XBMCKMOD_LALT) keyMapTable = keyMapTable|K_ALTLTAB; 
++      if (m_kbMods & XBMCKMOD_LSHIFT) keyMapTable = keyMapTable|K_SHIFTLTAB;
++      if (m_kbMods & XBMCKMOD_RSHIFT) keyMapTable = keyMapTable|K_SHIFTRTAB;
++      if (m_kbMods & XBMCKMOD_LCTRL) keyMapTable = keyMapTable|K_CTRLLTAB;
++      if (m_kbMods & XBMCKMOD_RCTRL) keyMapTable = keyMapTable|K_CTRLRTAB;
++    }
++
++    CLog::Log(LOGDEBUG, "CLinuxInputDevice::KeyEvent: keyMods 0x%04x kbMods 0x%04x keyMapTable 0x%04x", m_keyMods, m_kbMods, keyMapTable);
++
++    unsigned short keyMapValue = GetKeymapEntry(keyMapTable, code);
+ 
+-      if (keyMapValue != XBMCK_UNKNOWN)
++    /* the return value should be either a usable ascii code (< 256), a unicode value > 256
++     *  or a linux keyboard message which needs interpretation
++     *  */
++    //CLog::Log(LOGDEBUG, "CLinuxInputDevice::KeyEvent: keymapvalue 0x%04x returned from linux keymap", keyMapValue);
++
++    if (useKeyboard) // && keyMapValue > XBMCK_UNKNOWN)
++    {
++      // if there is an error, re set the kbMods (probably changed soft layouts)
++      if ( keyMapValue == K_HOLE || keyMapValue == K_NOSUCHMAP)
++        m_kbMods = 0;
++      
++      // check if the shift state has changed and prefer this to kodi's interpretation
++      if (KTYP(keyMapValue) == KT_LOCK  && devt.key.type == XBMC_KEYDOWN 
++          && (! (keyMapTable & K_SHIFTTAB) || m_keyMods & XBMCKMOD_SHIFT)) // eg shift-alt but not shiftlock-alt
+       {
+-        devt.key.keysym.sym = (XBMCKey) keyMapValue;
+-        if (keyMapValue > 0 && keyMapValue < 127)
++        m_kbMods = UpdateKeyboardModifiers(keyMapValue);
++        devt.key.keysym.sym = XBMCK_UNKNOWN; //NFA
++        CLog::Log(LOGDEBUG, "CLinuxInputDevice::KeyEvent: kb modifier changed to 0x%04x and keyMods is 0x%04x", m_kbMods, m_keyMods);
++      }
++
++      else
++      {
++        devt.key.keysym.mod = UpdateModifiers(devt);
++        if (keyMapValue & 0x8000) // assume unicode (utf-16)
++        {
++          devt.key.keysym.unicode = keyMapValue ^ 0xf000;
++          // if we have unicode, use it and stop kodi using sym
++          devt.key.keysym.sym = XBMCK_UNKNOWN;
++          CLog::Log(LOGDEBUG, "CLinuxInputDevice::KeyEvent: using 0x%04x as unicode", keyMapValue);
++        }
++        // use the extended character set but only if an 'ascii' key was pressed
++        else if (keyMapValue < 256 && keyMapValue > 0x20 && devt.key.keysym.sym < 128)
```

Comment:
To be able to press space in keyboard mode should be changed to:
`else if (keyMapValue < 256 && keyMapValue >= 0x20 && devt.key.keysym.sym < 128)`